nupic/research/frameworks/vernon/experiments/base_experiment.py

```diff
@@ -0,0 +1,184 @@
+
+            - local_dir: Results path
+            - logdir: Directory generated by Ray Tune for this Trial
+            - log_timestep_freq: Configures mixins and subclasses that log every
+                                 timestep to only log every nth timestep (in
+                                 addition to the final timestep of each epoch).
+                                 Set to 0 to log only at the end of each epoch.
+            - progress: Show progress during training
+            - name: Experiment name. Used as logger name
+            - log_level: Python Logging level
+            - log_format: Python Logging format
+            - seed: the seed to be used for pytorch, python, and numpy
+            - checkpoint_at_init: boolean argument for whether to create a checkpoint
+                                  of the initialized model. this differs from
+                                  `checkpoint_at_start` for which the checkpoint occurs
+                                  after the first epoch of training as opposed to
+                                  before it
+            - launch_time: time the config was created (via time.time). Used to report
+                           wall clock time until the first batch is done.
+                           Default: time.time() in this setup_experiment().
+        """
+        # Configure logging related stuff
+        log_format = config.get("log_format", logging.BASIC_FORMAT)
+        log_level = getattr(logging, config.get("log_level", "INFO").upper())
+        console = logging.StreamHandler()
+        console.setFormatter(logging.Formatter(log_format))
+        self.logger = logging.getLogger(config.get("name", type(self).__name__))
+        self.logger.setLevel(log_level)
+        self.logger.addHandler(console)
+        self.progress = config.get("progress", False)
+        self.launch_time = config.get("launch_time", time.time())
+        self.logdir = config.get("logdir", None)
+
+        if config.get("disable_logger", False):
+            self.logger.disabled = True
+            self.progress = False
+
+        # Configure seed
+        self.seed = config.get("seed", 42)
+        set_random_seed(self.seed, False)
+
+        self.logger.info("Execution order: %s",
+                         pformat(self.get_execution_order()))
+
+    @abc.abstractmethod
+    def run_iteration(self):
+        """
+        Run one training iteration of the experiment and return some sort of
+        result.
+        """
+
+    @abc.abstractmethod
+    def should_stop(self):
+        """
+        Whether or not the experiment should stop. Usually determined by the
+        number of epochs but customizable to any other stopping criteria
+        """
+
+    def stop_experiment(self):
+        """
+        Perform any needed cleanup.
+        """
+
+    def run_pre_experiment(self):
+        """
+        Evaluate the model before performing any training and return some sort
```

Comment:
Generally, the pre_experiment could be anything. For instance, one could record the sparsity level of the model at every epoch.